### PR TITLE
Use a unique fingerprint for a resource that is consistent after caching.

### LINKED ISSUE #10147: Missing intent in domain breaks UnexpecTED intent policy
### Rasa Open Source version

3.0.0rc1

### Rasa SDK version

_No response_

### Rasa X version

_No response_

### Python version

3.8

### What operating system are you using?

OSX

### What happened?

When I have an intent in my NLU data which is not in the domain, having a prediction with this intent in the tracker then breaks unexpecTED. Probably a similar a case as if we’d have a tracker with old intents.

```
Traceback (most recent call last):
  File "/Users/tayfun/.virtualenvs/rasa-rc-38/lib/python3.8/site-packages/rasa/engine/graph.py", line 455, in __call__
    output = self._fn(self._component, **run_kwargs)
  File "/Users/tayfun/.virtualenvs/rasa-rc-38/lib/python3.8/site-packages/rasa/core/policies/unexpected_intent_policy.py", line 615, in predict_action_probabilities
    is_unlikely_intent = self._check_unlikely_intent(
  File "/Users/tayfun/.virtualenvs/rasa-rc-38/lib/python3.8/site-packages/rasa/core/policies/unexpected_intent_policy.py", line 718, in _check_unlikely_intent
    predicted_intent_scores = {
  File "/Users/tayfun/.virtualenvs/rasa-rc-38/lib/python3.8/site-packages/rasa/core/policies/unexpected_intent_policy.py", line 719, in <dictcomp>
    index: similarities[0][index] for index, intent in enumerate(domain.intents)
IndexError: index 12 is out of bounds for axis 0 with size 12
```

### Command / Request

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by joejuzl**
Use a unique fingerprint for a resource that is consistent after caching. This ensures we always retrieve the correct resource for a node from the cache.

Fixes #10147

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by joejuzl** [comment]
I wanted to also add a more realistic test to `test_model_trainining.py` but it was taking too long and we need to get this merged today - so I think this is sufficient for now.

File: rasa/engine/caching.py [hunk 1]
```diff
@@ -434,6 +445,7 @@ def _load_from_cache(
         result_type: Text,
         node_name: Text,
         model_storage: ModelStorage,
+        output_fingerprint_key,
```
Comment (wochinge) [inline rasa/engine/caching.py:448]:
```suggestion
        output_fingerprint_key: Text,
```

File: tests/engine/training/test_graph_trainer.py [hunk 2]
```diff
@@ -494,6 +494,118 @@ def test_graph_trainer_train_logging_with_cached_components(
...
+    # cache.
+    train_with_schema(train_schema, temp_cache)
+
+    with temp_cache._sessionmaker.begin() as session:
+        # This will get the new cache entry for the "process" node.
+        query_for_most_recently_used_entry = sa.select(temp_cache.CacheEntry).order_by(
+            temp_cache.CacheEntry.last_used.desc()
+        )
+        entry = session.execute(query_for_most_recently_used_entry).scalars().first()
+        # Assert the fingerptint key of the new entry is the same. This confirms that
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:604]:
```suggestion
        # Assert the fingerprint key of the new entry is the same. This confirms that
```

**Review by wochinge — COMMENTED** [review]
Thanks for uncovering this huge bug and fixing it 💯 Just some typos and questions re the tests

File: tests/engine/training/test_graph_trainer.py [hunk 3]
```diff
@@ -494,6 +494,118 @@ def test_graph_trainer_train_logging_with_cached_components(
...
+    # returns 4 whereas it should be the second resource which returns 5, and the schema
+    # assert_node expects 5 now.
+    train_with_schema(train_schema, temp_cache)
+
+
+def test_resources_fingerprints_remain_after_being_cached(
+    tmp_path: Path,
+    temp_cache: LocalTrainingCache,
+    train_with_schema: Callable,
+    caplog: LogCaptureFixture,
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:549]:
```suggestion
```

File: tests/engine/training/test_graph_trainer.py [hunk 4]
```diff
@@ -494,6 +494,118 @@ def test_graph_trainer_train_logging_with_cached_components(
         }
 
 
+def test_resources_fingerprints_are_unique_when_cached(
+    tmp_path: Path,
+    temp_cache: LocalTrainingCache,
+    train_with_schema: Callable,
+    caplog: LogCaptureFixture,
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:501]:
```suggestion
```

File: tests/engine/training/test_graph_trainer.py [hunk 5]
```diff
@@ -494,6 +494,118 @@ def test_graph_trainer_train_logging_with_cached_components(
         }
 
 
+def test_resources_fingerprints_are_unique_when_cached(
+    tmp_path: Path,
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:498]:
```suggestion
```

File: tests/engine/training/test_graph_trainer.py [hunk 6]
```diff
@@ -494,6 +494,118 @@ def test_graph_trainer_train_logging_with_cached_components(
...
+
+    # This breaks when `Resource`s use the node name as a fingerprint.
+    # This is because the `Resource` for the first run is retrieved from the cache which
+    # returns 4 whereas it should be the second resource which returns 5, and the schema
+    # assert_node expects 5 now.
+    train_with_schema(train_schema, temp_cache)
+
+
+def test_resources_fingerprints_remain_after_being_cached(
+    tmp_path: Path,
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:546]:
```suggestion
```

File: tests/engine/training/test_graph_trainer.py [hunk 7]
```diff
@@ -494,6 +494,112 @@ def test_graph_trainer_train_logging_with_cached_components(
...
+            "process": SchemaNode(
+                needs={"resource": "train"},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={},
+            ),
+            "assert_node": SchemaNode(
+                needs={"i": "process"},
+                uses=AssertComponent,
```
Comment (wochinge) [inline tests/engine/training/test_graph_trainer.py:519]: I think the issue is that this one returns a `CacheableText`. If this component returns something non cacheable then the test works

Comment (joejuzl) [reply tests/engine/training/test_graph_trainer.py:519]: good catch

**Review by wochinge — APPROVED** [review]
🚀 Phew. Good that you caught this one - that would have been very bad 🙈

**Pull request merged**